### PR TITLE
Snyk extension implementation

### PR DESCRIPTION
**Pull request by tjkandala**
<details open>
<summary>JS monorepo with vulnerabilities in multiple projects</summary>

</details>

<details>
<summary>Java monorepo with no vulnerabilities</summary>

</details>

Limitations:
- Snyk only monitors the [default branch](https://support.snyk.io/hc/en-us/articles/360001497578-Which-branch-does-Snyk-Monitor), and they don't return the commit ID at the time they tested the project, so there's no way to guarantee that line decorations are accurate. Would it still be worth decorating import statements, even if there's a chance that the dependency in question is actually of a safe version at the viewed commit? I kept the logic for merged language patterns commented in case we decide to move forward with line decorations (also lives on in [sourcegraph/npm-audit](https://github.com/sourcegraph/sourcegraph-npm-audit/pull/1), so implementation wasn't a total waste).
- See [comment](https://github.com/sourcegraph/sourcegraph-snyk/pull/2#discussion_r510546384)
- Panel links for older alerts are useless, as the panel now contains the report from the latest opened file. 

TODO:
- README pictures once the panel UI is finalized

### CODE REVIEW

File: src/snyk.ts [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+                        // Check if Snyk has info for this project. We need this step because shortRepoName is not a valid project ID
+                        const allProjects = await listAllProjects({ orgId, options: apiOptions })
+
+                        // Get the closest project to this file in this repo
+                        const closestProject = getClosestProject(shortRepoName, filePath, allProjects.projects)
+
+                        // Problems in big monorepos:
+                        // - Could cause false positives for files of languages unsupported by Snyk: the closest
+                        // manifest file would be for a different language.
```
Comment (tjkandala) [inline src/snyk.ts]: Linked from "Limitations"

File: src/snyk.ts [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    return markdownString
+}
+
+export function getClosestProject(shortRepoName: string, filePath: string, allProjects: Project[]): Project {
+    // filter out projects wihout same repo name
+    const projects = allProjects.filter(({ name }) => name.includes(shortRepoName))
+
+    // get manifest file paths (maintain reference to project somehow.. id?)
```
Comment (tjkandala) [inline src/snyk.ts]: outdated comments (pre impl), will remove

File: src/snyk.ts [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+                            ),
+                            apiToken,
+                        }
+
+                        panelView.content = 'Loading...'
+                        const uri = new URL(editor.document.uri)
+                        const shortRepoName = decodeURIComponent(uri.pathname).replace(/^\//, '')
+                        const filePath = decodeURIComponent(uri.hash.slice(1))
+
+                        const repositoryNamePattern = '(?:^|/)([^/]+)/([^/]+)$'
```
Comment (marekweb) [inline src/snyk.ts]: **question** Were you intending to get this value from the setting `snyk.repositoryNamePattern`?

**Review by felixfbecker — APPROVED** [review]
This is an impressive demonstration of the flows that are possible with the extension API!

UX-wise for Snyk in particular, I wonder if notifications could be annoying. The vulnerability reports are unfortunately often not that relevant (more noise) so I would be careful with UI that shows the user notifications vs something "inline". 
Given this UI is more repo-level, maybe the panel content would actually make more sense as a view provider on the repo homepage? They don't have to return graphs, they can return markdown too. It would require the `codeInsights` feature flag though.

Given that and that there are still a lot of TODOs in the code and that it's Friday, I'd propose we merge this as is, but keep the `"wip": true` and come back to it later. Only thing I'd update before merging to not make the extension registry messy is updating the description in package.json, adding an icon to package.json and adding the description (one sentence) to the README too.

**Review by marekweb — APPROVED** [review]

File: src/snyk.ts (hunk 3 above)
Comment (tjkandala) [reply src/snyk.ts]: Yes, thanks!

**Pull request merged**